File: src/features/typing/data/java/maximum_product_subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int best = nums[0];
        int maxP = nums[0];
        int minP = nums[0];
        for (int i = 1; i < nums.length; i++) {
            int num = nums[i];
            if (num < 0) {
                int temp = maxP;
                maxP = minP;
                minP = temp;
            }
            maxP = Math.max(num, maxP * num);
            minP = Math.min(num, minP * num);
            if (maxP > best) {
                best = maxP;
            }
        }
        return best;
    }
}

File: src/features/typing/data/dart/maximum_product_subarray.dart
class Solution {
  int maxProduct(List<int> nums) {
    var best = nums[0];
    var maxP = nums[0];
    var minP = nums[0];
    for (var i = 1; i < nums.length; i++) {
      final num = nums[i];
      if (num < 0) {
        final temp = maxP;
        maxP = minP;
        minP = temp;
      }
      maxP = num > maxP * num ? num : maxP * num;
      minP = num < minP * num ? num : minP * num;
      if (maxP > best) {
        best = maxP;
      }
    }
    return best;
  }
}
